### PR TITLE
Fix xHamster support, fixes #264

### LINKED ISSUE #264: How to use referer option?
# New issue

## Your environment

Please answer these questions when reporting a new issue:

**What is your operating system (Windows, Linux, OSX, etc.)?** Debian

**What is your web server (Apache, IIS, etc.)?** Apache

**What version of AllTube are you using?** 2.1.0

**How did you install AllTube (with Git or with a release package)?** Composer

**What version of PHP are you using?** 7.3.15

**What version of Python are you using?** 2.7.13

**What version of youtube-dl are you using?** 2020.02.16

**Do you get any PHP-related errors in your webserver's logs?** Invalid filter specification u'referer=https://xhamster.com'

**What is the content of your `config/config.yml` file?** No file

**Please provide the URL of a video that causes the issue.** Any video of xhamster.com

## Describe your issue

Hello,
Youtube-dl has problems with xhamster.com. It shows 403 error for any video. But it works with option --referer='https://xhamster.com'

youtube-dl -f '[height<=720]' --referer='https://xhamster.com' https://xhamster.com/....... (any video)

How can I use this option in alltube?

$video = new Video($url, [protocol=https][height<=720]);
$video_url = $video->getUrl();

### PR DESCRIPTION
**Pull request by 0x6470**
As described in #264 xHamster downloads are not possible, unless their site itself is specified as HTTP-Referer.
Since it didn't work for me either, I tried to fix it.
As it seems, it only works if the referer is passed as custom stream context option, as this overrides other values. 

further details here:
<http://docs.guzzlephp.org/en/stable/faq.html#how-can-i-add-custom-stream-context-options>
<https://www.php.net/manual/en/context.http.php#context.http.header>

Setting the referer as Guzzle request header (<http://docs.guzzlephp.org/en/stable/request-options.html#headers>) or through CURLOPT_REFERER did not work.

With this fix, it's possible to download videos from xHamster if the stream option is activated.
To achieve this without the streaming option, I recommend to use a plugin in your web browser.

### CODE REVIEW

File: classes/Video.php [hunk 1]
```diff
@@ -613,12 +613,22 @@ public function getHttpResponse(array $headers = [])
     {
         $client = new Client();
         $urls = $this->getUrl();
+        $stream_context_options = [];
+
+        if (strpos($this->webpageUrl, 'xhamster.com') !== false) {
```
Comment (Rudloff) [inline classes/Video.php:618]: I would prefer a more generic solution (other websites might require a referer).
You can check if there is a `Referer` key in `$this->http_headers` and reuse that.

Comment (0x6470) [reply classes/Video.php]: I understand your concerns.
Something like this should work in a more general way.

File: classes/Video.php [hunk 2]
```diff
@@ -613,12 +613,22 @@ public function getHttpResponse(array $headers = [])
     {
         $client = new Client();
         $urls = $this->getUrl();
+        $stream_context_options = [];
+
+        if (array_key_exists('Referer', $this->http_headers)) {
```
Comment (Rudloff) [inline classes/Video.php:618]: `http_headers` is an object so you should explicitly convert it to an array:
```suggestion
        if (array_key_exists('Referer', (array)$this->http_headers)) {
```

Comment (0x6470) [reply classes/Video.php]: Thanks again!
I did the small adjustment.

**Comment by Rudloff** [comment]
I reverted this change because it was overriding the other headers.

Guzzle already takes care of creating the context [here](https://github.com/guzzle/guzzle/blob/e8ed4dbf49b260ff129ff0e0400718c3269971bf/src/Handler/StreamHandler.php#L374).

**Pull request merged**